### ISSUE TITLE
Optimize/refine NFI sparse matrix method

### ISSUE THREAD

**Issue by courtneypeck** [punch-mission/punchbowl#1077]
Now that we have implemented the sparse matrix algorithm into our code, we can start to optimize the free parameters of the method. Work in this ticket will note the free parameters to tune and begin testing the impact of these parameters on the NFI dynamic stray light removal. Additional tickets should be opened as the work evolves.

**Comment by Rokkaan5** [comment]
# Main todo list
Determine the following

- [x] fine-tuneable parameters of the matrix method 
- [x] Open tickets to fine-tune (grouped) parameters as see fit

## Stretch goals

- [x] rank the impacts of tunable parameters (e.g. high impact parameters vs. fine-tuning)
- [ ] Understand the reasons for the default values
- [ ] Understand how the parameters affect results (may end up being a separate ticket once tunable parameters are determined)

**Comment by Rokkaan5** [comment]
# 7/29 notes

Based on my general knowledge of the code at this point, I knew the high level tunable parameters should mostly be associated with the **forward matrices** (i.e. kernels) and **solution reconstruction** (i.e. intensity profile).

The following video from Sam's presentation on the topic helps provide context and understanding
https://github.com/user-attachments/assets/df6194ea-7f5b-4da1-9f55-86f0ac46b0be

## Today's deep dive on the tunable parameters for the generating the forward matrices (kernels):
`inst_forward_mat` and `sky_forward_mat` are mainly made from the internal `element_source_responses()` function in `nfi_modules`. (Which I didn't look deeply at the parameter impacts yet, but maybe I should come back to eventually.)

And `stray_forward_mat` (i.e. the forward matrix for the stray light model) creates the list of counter-clockwise rotating kernels mainly from the `generate_kernels()` function (from the `straylight_kernels.py` script in `nfi_modules`).

### At a high level 
The main tunable parameters of `generate_kernels()` (i.e. the stray light model kernels) can fall into three categories:

- The **size** of the kernels (which also stays constant throughout the entire rotation)
- The **intensity gradient** of the kernel(s)
- The **rotation parameters** of the kernels (center of rotation and elongation angle of kernel w.r.t. image center)

Jasmine's sloppy notes on the categorized parameters:
<img width="468" height="280" alt="Image" src="https://github.com/user-attachments/assets/9e24b9b4-7869-4d84-829a-7682dbf84fda" />


### Extra (optional) note
All forward matrices have a final step to be reformatted in the following way
<img width="514" height="150" alt="Image" src="https://github.com/user-attachments/assets/9b1cd6dd-5433-4dc1-af37-540eec69fd0d" />

Where "X" substitutes for "inst", "sky", or "stray" (only color coded to visually distinguish norms vs the initial forward matrix). Since it's applied to all forward matrices, not potentially relevant for tunable parameter concern, but just sparks interest as to what final matrix looks like.

**Comment by Rokkaan5** [comment]
No detailed notes this time, but from what I could tell so far:

## Intensity-Profile tunable parameters
there's a lot more parameters that go into generating the intensity profile.
At a generally high-level, most parameters seem to be generally grouped into the following sub-categories:

- Forward function (non-linear mapping)
- Regularization ("loss function") tuning
- Chi-squared convergence tuning


## List of main top-level tunable parameters (somewhat ranked by impact)

- Dynamic stray light Intensity profile parameters
- Sky and Instrument stray light regularization
- Dynamic Stray light kernel parameters
- Glint masking (the size of the masks and where they're centered)
- (I don't _think_ the sky and instrument kernels are tunable?)

**Comment by Rokkaan5** [comment]
# Jasmine's final remarks

Issues punch-mission/sdf#22 and punch-mission/sdf#23 address the sub-tasks to explore and refine the parameters for dynamic light intensity profile and kernel generation respectively.
(I originally had them assigned to punchbowl, but since I designed them more as tasks to be explored via jupyter notebooks, etc., I moved them to the sdf repo)

As in my last note for this ticket, I wanted to keep note of the fact that technically how the background starfield and f-corona (categorized as "sky" in the matrix method) gets processed in this matrix method also will have some impact on the final processed image. But before I make any defined ticket for it, I would like to communicate with @svank (after he's back from his vacation and SPD) regarding some ideas/thoughts on this, as I recall from a meeting that he also had some ideas related to this (e.g. have the matrix method work with data products with the starfield already removed so this matrix method doesn't have to worry about it in its own fitting). I have in my own personal to-do list to set up a meeting regarding this.

Lastly, I don't think it's worth making a separate ticket for "fine-tuning" the glint masking parameters. While it _technically_ will affect the final image (by how much you mask out and etc.). I mainly left it as one of the parameters for thorough notes purposes. But I think the default parameters for the masking dimensions are good enough for the current standards, and I also think any fine-tuning related to this will get naturally resolved by the tasks above or perhaps become irrelevant altogether.

No code was generated or modified to find these notes, so there's no associated PR for this issue.

Based on the notes above, I think this ticket can be considered resolved for the time being (and the overarching goals are to be carried on by the tickets created by this issue).
